[JSC] Adjust max length of wasm resizable buffer on 32-bit

mikhailramalho opened the pull request:
#### 09f1728830bec3f5b3b39b910ee62fd153c8a875
<pre>
[JSC] Adjust max length of wasm resizable buffer on 32-bit
<a href="https://bugs.webkit.org/show_bug.cgi?id=298045">https://bugs.webkit.org/show_bug.cgi?id=298045</a>

Reviewed by Justin Michaud.

It&apos;s now 2GB on 32-bit jsc. Test also updated. 64-bit is NFC.

* JSTests/wasm/js-api/memory-toResizableBuffer.js:
(assertSharedGrowableBufferOfPageSize):
* Source/JavaScriptCore/wasm/js/JSWebAssemblyMemory.cpp:
(JSC::JSWebAssemblyMemory::associateArrayBuffer):

Canonical link: <a href="https://commits.webkit.org/299400@main">https://commits.webkit.org/299400@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/234ac18ad5cd21dde69267dcee0f0b6fb3b02298

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/118768 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/38449 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/29100 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/124953 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/70829 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/39145 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/47031 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/90140 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/59650 "Found 1 new test failure: workers/btoa-oom.html (failure)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/121721 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31189 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/106479 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70646 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/30249 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/24592 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/68612 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/110886 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/100632 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/24780 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/128007 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/117282 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/45675 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/34479 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/98787 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/46039 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/102699 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98569 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25079 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/44009 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/22014 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/42205 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/45545 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/51223 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/145978 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/45009 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/37538 "Found 1 new JSC binary failure: testapi, Found 18598 new JSC stress test failures: ChakraCore.yaml/ChakraCore/test/Array/array_slice.js.default, ChakraCore.yaml/ChakraCore/test/Array/array_sort2.js.default, ChakraCore.yaml/ChakraCore/test/Array/push2.js.default, ChakraCore.yaml/ChakraCore/test/Array/reverse1.js.default, ChakraCore.yaml/ChakraCore/test/Array/toLocaleString.js.default, ChakraCore.yaml/ChakraCore/test/Bugs/blue_1086262.js.default, ChakraCore.yaml/ChakraCore/test/Date/date_cache_bug.js.default, ChakraCore.yaml/ChakraCore/test/Error/CallNonFunction.js.default, ChakraCore.yaml/ChakraCore/test/Error/stack2.js.default, ChakraCore.yaml/ChakraCore/test/Function/FuncBody.bug227901.js.default ... (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/48355 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/46695 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

ast-hugger commented on Source/JavaScriptCore/wasm/js/JSWebAssemblyMemory.cpp:
```diff
@@ -87,13 +87,11 @@ void JSWebAssemblyMemory::associateArrayBuffer(JSGlobalObject* globalObject, boo
-#endif
+            // If sizeof(size_t) == 4, we say it's 2^15 pages which is not right but at least it still builds for ARMv7.
+            const size_t defaultMaxByteLengthIfMemoryHasNoMax = static_cast<size_t>(PageCount::fromBytes(MAX_ARRAY_BUFFER_SIZE).bytes());
+
```
This should stay conditional, with the baseline 64-bit case not dependent on `MAX_ARRAY_BUFFER_SIZE`. The default maximum size of `65536 * 65536` is dictated by the spec (https://webassembly.github.io/threads/js-api/index.html#create-a-resizable-memory-buffer, step 4.1). It happens to match our `MAX_ARRAY_BUFFER_SIZE`, but if we increased it, `defaultMaxByteLengthIfMemoryHasNoMax` would have to stay unchanged to pass the tests.

mikhailramalho replied:
understood, reverted.

ast-hugger commented on Source/JavaScriptCore/runtime/PageCount.h:
```diff
@@ -42,7 +42,7 @@ static_assert(sizeof(size_t) == sizeof(uint64_t));
 // Because we are using a size_t to store the size in bytes of array buffers, we cannot support 4GB on 32-bit platforms.
 // So we are sticking with 2GB. It should in theory be possible to support up to (4GB - 1B) if anyone cares.
-#define MAX_ARRAY_BUFFER_SIZE 0x7fffffffu
+#define MAX_ARRAY_BUFFER_SIZE 0x80000000u
```
I'll let someone else review and comment on this. This change is not strictly necessary for this PR, and perhaps there was a motivation to not make the limit an even `2^15`.

Constellation replied:
Basically you need to enable LARGE_TYPED_ARRAY feature. In these platforms, ArrayBuffer max size is 0x7ffffffu, because ArrayBuffer is holding the size in `size_t`, which is 32bit in these platforms. Increasing this will become a security issue if you are also not enabling `LARGE_TYPED_ARRAY`.
So given that this code path is for `!USE(LARGE_TYPED_ARRAY)`, this change is not correct.

Constellation replied:
If you would like to allow the size more than 0x7fffffff, changing this code here is not the right direction. You need to enable `USE(LARGE_TYPED_ARRAY)` and changing all the field storing size_t for TypedArray / ArrayBuffer, JIT code access / interpreter access, SerializedScriptValue, and so on. So kind of very large project.

Constellation replied:
Particular reason why we cannot use 0x80000000 here is because Baseline / DFG / FTL is saying that result of TypedArray#length will be Int32, so 0x8000000 becomes negative number. `USE(LARGE_TYPED_ARRAY)` code path is handling them correctly, but assuming that size is 64bit. So anyway, just changing this is not enough.

mikhailramalho replied:
Thanks for the explanation, I mistakenly thought this was ever used as a `size_t` for array lengths.

mikhailramalho replied:
Unfortunately, we can't set `defaultMaxByteLengthIfMemoryHasNoMax` to `0x7fffffffu` or JSC aborts with:
```
ASSERTION FAILED: bytes % pageSize == 0
/Debug/JavaScriptCore/PrivateHeaders/JavaScriptCore/PageCount.h(77) : static JSC::PageCount JSC::PageCount::fromBytes(uint64_t)
```
But since it's okay to change the test case, I reverted the changes to `JSWebAssemblyMemory.cpp` and only changed the test case to do the right calculation.

ast-hugger commented on JSTests/wasm/js-api/memory-toResizableBuffer.js:
```diff
@@ -111,5 +111,6 @@ function assertSharedGrowableBufferOfPageSize(pageCount, buffer, maxPageCount) {
     let buffer = memory.toResizableBuffer();
-    assertEq(buffer.maxByteLength, pageSize * 65536);
+    const expectedPageCount = is32BitPlatform() ? 65535 : 65536;
+    assertEq(buffer.maxByteLength, pageSize * expectedPageCount);
```
Sorry, I probably wasn't very clear on the complete picture. We shouldn't simply use 2^16-1 pages on 32 bits like I originally did, because my intention was just to make it buildable. Instead we should:

- In the 32-bit `#if` branch in `JSWebAssemblyMemory::associateArrayBuffer`, define `defaultMaxByteLengthIfMemoryHasNoMax` to be the largest value that fits within `MAX_ARRAY_BUFFER_SIZE` and contains an integral number of pages. That's what `PageCount::fromBytes(MAX_ARRAY_BUFFER_SIZE).bytes()` that I suggested would compute.
- That number of pages is 2^15-1, or 32767. That's the number we should expect in this test in the 32-bit case.

As Yusuke pointed out, `MAX_ARRAY_BUFFER_SIZE` in the 32-bit case is deliberately set to what it is, and we shouldn't touch it.

mikhailramalho replied:
done

justinmichaud reviewed (approved):
We should not enable this feature on armv7 if it does not follow the spec

mikhailramalho committed:
[JSC] Adjust max length of wasm resizable buffer on 32-bit
https://bugs.webkit.org/show_bug.cgi?id=298045

Reviewed by Justin Michaud.

It's now 2GB on 32-bit jsc. Test also updated. 64-bit is NFC.

* JSTests/wasm/js-api/memory-toResizableBuffer.js:
(assertSharedGrowableBufferOfPageSize):
* Source/JavaScriptCore/wasm/js/JSWebAssemblyMemory.cpp:
(JSC::JSWebAssemblyMemory::associateArrayBuffer):

Canonical link: https://commits.webkit.org/299400@main
